okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java

```diff
@@ -18,29 +18,47 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import okhttp3.HttpUrl;
 
 final class UrlSelectorImpl implements UrlSelector {
 
     private final ImmutableList<HttpUrl> baseUrls;
     private final AtomicInteger currentUrl;
+    private final Cache<HttpUrl, Boolean> failedUrls;
+    private final boolean useFailedUrlCache;
 
-    private UrlSelectorImpl(ImmutableList<HttpUrl> baseUrls) {
+    private UrlSelectorImpl(ImmutableList<HttpUrl> baseUrls, long duration, TimeUnit unit) {
         this.baseUrls = baseUrls;
         this.currentUrl = new AtomicInteger(0);
+        this.failedUrls = CacheBuilder.newBuilder()
+                .maximumSize(baseUrls.size())
+                .expireAfterWrite(duration, unit)
+                .build();
+        this.useFailedUrlCache = duration > 0L;
 
         Preconditions.checkArgument(!baseUrls.isEmpty(), "Must specify at least one URL");
+        Preconditions.checkArgument(duration >= 0L, "Cache expiration must be non-negative");
     }
 
-    static UrlSelectorImpl create(Collection<String> baseUrls, boolean randomizeOrder) {
+    /**
+     * Creates a new {@link UrlSelector} with the supplied URLs. The order of the URLs may be randomized by setting
+     * {@code randomizeOrder} to true. If a non-zero {@code duration} and {@code timeUnit} are specified, URLs that are
+     * marked as failed using {@link #markAsFailed(HttpUrl)} will be removed from the pool of prioritized, healthy URLs
+     * for that period of time.
+     */
+    static UrlSelectorImpl create(Collection<String> baseUrls, boolean randomizeOrder, long duration,
```

Comment:
We now have two methods called `create` which do different things.  Can we go for descriptive names?  e.g. 

```
create()
createWithFailedUrlCooloff()
```